crates/miden-lib/src/account/interface/mod.rs

```diff
@@ -0,0 +1,384 @@
+    sync::Arc,
+    vec::Vec,
+};
+
+use miden_objects::{
+    account::{Account, AccountCode, AccountId, AccountProcedureInfo, AccountType},
+    assembly::mast::{MastForest, MastNode, MastNodeId},
+    crypto::dsa::rpo_falcon512,
+    note::{Note, NoteScript},
+    Digest,
+};
+
+use crate::{
+    account::components::{
+        basic_fungible_faucet_library, basic_wallet_library, rpo_falcon_512_library,
+    },
+    note::scripts::{p2id_commitment, p2idr_commitment, swap_commitment},
+    AuthScheme,
+};
+
+#[cfg(test)]
+mod test;
+
+// ACCOUNT INTERFACE
+// ================================================================================================
+
+/// An [`AccountInterface`] describes the exported, callable procedures of an account.
+///
+/// A note script's compatibility with this interface can be inspected to check whether the note may
+/// result in a successful execution against this account.
+pub struct AccountInterface {
+    account_id: AccountId,
+    auth: Vec<AuthScheme>,
+    component_interfaces: BTreeSet<AccountComponentInterface>,
+}
+
+impl AccountInterface {
+    // CONSTRUCTORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Creates a new [`AccountInterface`] instance from the provided account ID, authentication
+    /// schemes and account code.
+    pub fn new(account_id: AccountId, auth: Vec<AuthScheme>, code: &AccountCode) -> Self {
+        let component_interfaces = AccountComponentInterface::from_procedures(code.procedures());
+
+        Self { account_id, auth, component_interfaces }
+    }
+
+    // PUBLIC ACCESSORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns a reference to the account ID.
+    pub fn id(&self) -> &AccountId {
+        &self.account_id
+    }
+
+    /// Returns the type of the reference account.
+    pub fn account_type(&self) -> AccountType {
+        self.account_id.account_type()
+    }
+
+    /// Returns true if the reference account can issue assets.
+    pub fn is_faucet(&self) -> bool {
+        self.account_id.is_faucet()
+    }
+
+    /// Returns true if the reference account is a regular.
+    pub fn is_regular_account(&self) -> bool {
+        self.account_id.is_regular_account()
+    }
+
+    /// Returns true if the reference account is public.
+    pub fn is_public(&self) -> bool {
+        self.account_id.is_public()
+    }
+
+    /// Returns a reference to the vector of used authentication schemes.
+    pub fn auth(&self) -> &Vec<AuthScheme> {
+        &self.auth
+    }
+
+    /// Returns a reference to the set of used component interfaces.
+    pub fn components(&self) -> &BTreeSet<AccountComponentInterface> {
+        &self.component_interfaces
+    }
+
+    /// checks if a note can be consumed against the current [AccountInterface] instance.
+    pub fn can_consume(&self, note: &Note) -> NoteAccountCompatibility {
```

Comment:
Should we rename this to be more aligned with what it actually expresses? Since we can't definitively say whether "a note can be consumed", I would make this answer the question "is this note compatible with this account interface?". So perhaps `is_compatible`, which would read like `account_interface.is_compatible(note)`.